sourmash_lib/lca/lca_utils.py

```diff
@@ -0,0 +1,302 @@
+"""
+Utility functions for lowest-common-ancestor analysis tools.
+"""
+from __future__ import print_function
+import sys
+import json
+import gzip
+from collections import OrderedDict, namedtuple, defaultdict, Counter
+
+try:                                      # py2/py3 compat
+    from itertools import zip_longest
+except ImportError:
+    from itertools import izip_longest as zip_longest
+import pprint
+
+import sourmash_lib._minhash
+from sourmash_lib.logging import notify
+
+# type to store an element in a taxonomic lineage
+LineagePair = namedtuple('LineagePair', ['rank', 'name'])
+
+
+# ordered list of taxonomic ranks
+def taxlist(include_strain=False):
+    for k in ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus',
+              'species']:
+        yield k
+    if include_strain:
+        yield 'strain'
+
+
+# produce an ordered list of tax names from lineage
+def zip_lineage(lineage, include_strain=False, truncate_empty=False):
+    """@CTB: document and test."""
+    row = []
+    empty = LineagePair(None, '')
+    for taxrank, lineage_tup in zip_longest(taxlist(), lineage,
+                                            fillvalue=empty):
+        debug('ABC', taxrank, lineage_tup, lineage_tup == empty)
+        if lineage_tup != empty and lineage_tup.name:
+            if lineage_tup.rank != taxrank:
+                raise ValueError('incomplete lineage at {}!? {}'.format(lineage_tup.rank, lineage))
+        else:
+            if truncate_empty:
+                break
+
+        row.append(lineage_tup.name)
+    return row
+
+
+# filter function toreplace blank/na/null with 'unassigned'
+filter_null = lambda x: 'unassigned' if x.strip() in \
+  ('[Blank]', 'na', 'null', '') else x
+null_names = set(['[Blank]', 'na', 'null'])
```

Comment:
make `null_names` into a constant (frozenset?), use it in `filter_null` (defined as a function, not as a lambda).
```python
NULL_NAMES = {'[Blank]', 'na', 'null', ''}

def filter_null(x):
    if x.strip() in NULL_NAMES:
        return 'unassigned'
    return x
```